File: src/net/liplum/api/cyber/IDataReceiver.java
package net.liplum.api.cyber;

import arc.struct.ObjectSet;
import mindustry.type.Item;
import net.liplum.CalledBySync;
import net.liplum.ClientOnly;
import net.liplum.lib.delegates.Delegate1;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public interface IDataReceiver extends IDataBuilding {
    void receiveData(@NotNull IDataSender sender, @NotNull Item item, int amount);

    /**
     * Gets the max acceptable number of this {@code item}.
     * -1 means any
     *
     * @param sender sender
     * @param itme   item
     * @return amount
     */
    int acceptedAmount(@NotNull IDataSender sender, @NotNull Item itme);

    /**
     * Gets what this receiver wants<br/>
     * null : Any<br/>
     * Array.Empty : Nothing<br/>
     * An array : what's in the array<br/>
     *
     * @return what this receiver wants
     */
    @Nullable
    Item[] getRequirements();

    @ClientOnly
    boolean isBlocked();

    @CalledBySync
    void connect(@NotNull IDataSender sender);

    @CalledBySync
    void disconnect(@NotNull IDataSender sender);

    @NotNull
    ObjectSet<Integer> connectedSenders();

    default boolean isConnectedWith(@NotNull IDataSender sender) {
        return connectedSenders().contains(sender.getBuilding().pos());
    }

    @NotNull
    Delegate1<IDataReceiver> getOnRequirementUpdated();

    @Nullable
    Integer connectedSender();

    /**
     * Gets the maximum limit of connection.<br/>
     * -1 : unlimited
     *
     * @return the maximum of connection
     */
    int maxSenderConnection();


    default boolean acceptConnection(@NotNull IDataSender host) {
        return canHaveMoreSenderConnection();
    }

    default boolean canHaveMoreSenderConnection() {
        int max = maxSenderConnection();
        if (max == -1) {
            return true;
        }
        return connectedSenders().size < max;
    }
}

File: main/src/net/liplum/api/cyber/IDataReceiver.kt
package net.liplum.api.cyber

import arc.graphics.Color
import arc.struct.ObjectSet
import arc.struct.Seq
import mindustry.type.Item
import net.liplum.R
import net.liplum.api.ICyberEntity
import net.liplum.common.Changed
import net.liplum.common.delegates.Delegate1
import net.liplum.mdt.CalledBySync
import net.liplum.mdt.ClientOnly

interface IDataReceiver : ICyberEntity {
    fun receiveDataFrom(sender: IDataSender, item: Item, amount: Int)
    /**
     * Gets the max acceptable number of this `item`.
     * -1 means any
     *
     * @param sender sender
     * @param item   item
     * @return amount
     */
    fun getAcceptedAmount(sender: IDataSender, item: Item): Int
    /**
     * Gets what this receiver wants<br></br>
     * null : Any<br></br>
     * Array.Empty : Nothing<br></br>
     * An seq : all things in the array<br></br>
     * Please cache this value, this is a mutable list.
     * @return what this receiver wants
     */
    val requirements: Seq<Item>?
    @ClientOnly
    val receiverColor: Color
        get() = R.C.Receiver
    @CalledBySync
    fun connectTo(sender: IDataSender) {
        connectedSenders.add(sender.building.pos())
    }
    @CalledBySync
    fun disconnectFrom(sender: IDataSender) {
        connectedSenders.remove(sender.building.pos())
    }

    val connectedSenders: ObjectSet<Int>
    fun isConnectedTo(sender: IDataSender): Boolean {
        return connectedSenders.contains(sender.building.pos())
    }

    val onRequirementUpdated: Delegate1<IDataReceiver>
    /**
     * Gets the maximum limit of connection.<br></br>
     * -1 : unlimited
     *
     * @return the maximum of connection
     */
    val maxSenderConnection: Int
    fun acceptConnectionTo(sender: IDataSender) = canHaveMoreSenderConnection
    val senderConnectionNumber: Int
        get() = connectedSenders.size
    val canHaveMoreSenderConnection: Boolean
        get() = maxSenderConnection == -1 || senderConnectionNumber < maxSenderConnection
}